File: src/scrape-lyrics/genius.R
library(tidyverse)
library(httr)
library(stringr)
library(rvest)

token <- '2PtHXnbMH-EAIrKUGitGgN2Edg1lmsLwjvU0jDaETseU8vCPdnxhV0AyslIW_vxD'
GET
genius_get_artists <- function(artist_name, n_results = 10) {
  baseURL <- 'https://api.genius.com/search?q=' 
  requestURL <- paste0(baseURL, gsub(' ', '%20', artist_name), '&per_page=', n_results, '&access_token=', token) #paste0=Concatenate vectors after converting to character.
  res <- GET(requestURL) %>% content %>% .$response %>% .$hits
  View(res)
  map_df(1:length(res), function(x) {
    tmp <- res[[x]]$result$primary_artist
    list( artist_id = tmp$id, artist_name = tmp$name )
  }) %>% unique
}

lyric_scraper <- function(url) {
    read_html(url) %>% 
    html_node('lyrics') %>% 
    html_text
}

genius_artists <- genius_get_artists('chvrches')
genius_artists

  baseURL <- 'https://api.genius.com/artists/' 
  requestURL <- paste0(baseURL, genius_artists$artist_id[1], '/songs')

  track_lyric_urls <- list()

  i <- 1
  while (i > 0) {
    tmp <- GET(requestURL, query = list(access_token = token, per_page = 50, page = i)) %>% content %>% .$response
    track_lyric_urls <- c(track_lyric_urls, tmp$songs)
    if (!is.null(tmp$next_page)) {
      i <- tmp$next_page
      } else {
        break
      }
  }

track_lyric_urls
  
genius_df <- map_df(1:length(track_lyric_urls), function(x) {
  lyrics <- lyric_scraper(track_lyric_urls[[x]]$url)
  tots <- list(track_name = track_lyric_urls[[x]]$title, lyrics = lyrics)
  return(tots)
})

write.csv(genius_df, file="chvrchesLyrics.csv")

File: src/scrape-lyrics/scrapelyrics/genius.py
import requests
import json
import bs4
import os
from collections import defaultdict

TOKEN = "2PtHXnbMH-EAIrKUGitGgN2Edg1lmsLwjvU0jDaETseU8vCPdnxhV0AyslIW_vxD"


def genius_get_artists(artist_name):
    """
    given an artist name as a string.
    returns the most frequent artist in the first 10 genius search results for that string.
    """
    try:
        URL = (
                "https://api.genius.com/search?q="
                + artist_name.replace(" ", "%20")
                + "&per_page=10"
        )
        raw = requests.get(URL, headers={"Authorization": "Bearer " + TOKEN})
        json_obj = json.loads(raw.content)
        result = defaultdict(int)
        for hit in json_obj["response"]["hits"]:
            hit = hit["result"]["primary_artist"]
            result[(hit["id"], hit["name"])] += 1
        return max(iter(result.items()), key=lambda x: x[1])[0]
    except:
        print("no results found. usage: genius_get_artist(string) return (id, name)")
        raise


def genius_get_songs(artist_id):
    """
    given an artist id, compile a list of songs as tuples (title, url)
    """
    try:
        artist_id = str(artist_id)
        result = []
        baseURL = "https://api.genius.com/artists/" + artist_id + "/songs"
        i = 1
        while i:
            URL = baseURL + "?per_page=50&page=" + str(i)
            raw = requests.get(URL, headers={"Authorization": "Bearer " + TOKEN})
            json_obj = json.loads(raw.content)
            for song in json_obj["response"]["songs"]:
                result.append((song["title"], song["url"]))
            next_i = json_obj["response"]["next_page"]
            print("next page = ", next_i)
            if next_i:
                print(len(result), "found, next page")
                i += 1
            else:
                print("done ", len(result), "songs found")
                break
        return result
    except:
        print(
            "no songs found. usage: genius_get_songs(id) return list of tuples [(title,url)...]"
        )
        raise


def genius_get_lyrics(song_url):
    """
    given a URL of a song, scrape the lyrics and return them as a unicode string
    """
    try:
        response = requests.get(song_url)
        soup = bs4.BeautifulSoup(response.text, "lxml")
        return soup.find("lyrics").text.strip()
    except:
        print(
            "cannot find lyrics. usage: genius_get_lyrics(song_url) returns lyrics as unicode string"
        )
        raise


def genius_get_all_lyrics_for_artist(artist):
    """
    given an artist, find the proper spelling, then compile a list of songs, then save lyrics to a file and report the progress
    """
    art = genius_get_artists(artist)
    song_list = genius_get_songs(art[0])

    try:
        if not os.path.exists("/home/thom/projects/Lyrics/" + art[1]):
            os.makedirs("/home/thom/projects/Lyrics/" + art[1])
        print("get lyrics for artist " + art[1])
        count = 0
        for song in song_list:
            lyricFile = open(
                "/home/thom/projects/Lyrics/"
                + art[1]
                + "/"
                + song[0].replace("/", "_")
                + ".txt",
                "wb",
            )
            surl = song[1]
            l = "\nTitle:" + song[0] + "\n" + genius_get_lyrics(surl)
            lyricFile.write(l.encode("utf-8", "ignore"))
            count += 1
            lyricFile.close()
    except:
        print(
            "cannot get lyrics usage: genius_get_all_lyrics_for_artist(artist) creates one text file per song retrieved from genius.com"
        )
        raise
